### PR TITLE
Timers

### PR DESCRIPTION
**Pull request by mrichards42**
This is basically a generalization of the idea from #46. I've been using the read_input timeout for timers in puzzles, but my timer code is a little scattered, and once I added a second timer the [logic got kind of hairy](https://github.com/mrichards42/remarkable_puzzles/blob/819a777e14c396689ddd87827294633489ebb8d8/src/main.cpp#L72-L75).

This should be way more lightweight than using a thread for timers, so I was also thinking about using this for longpress like we were discussing in https://github.com/rmkit-dev/rmkit/pull/87#issuecomment-774780011.

I changed animation_demo to use a timer instead of a thread as a test. The code doesn't actually look that different, but when I tested it behaved very differently under stress. Both the thread build and the timer build act pretty much the same with a single circle. With multiple circles the thread version's frame rate drops significantly. The timer version seems to keep the same frame rate, but the quality drops (the circles start to appear jagged). Makes sense that there's some tradeoff. The other big difference is that the thread version uses almost 3x more memory in my tests (~9% compared to ~3%).

Very open to feedback on this. I'm not set on any of the function names, or hierarchy, or the lack of namespace or anything if you have thoughts. I didn't add `set_timeout` or any variation to MainLoop (or even just as a plain ui:: function), but it might be a nicer abstraction if you don't have to reach all the way in to `ui::MainLoop::timers`.

### CODE REVIEW

**Review by raisjn — APPROVED** [review]
tumblr_lxm0d9LTWn1qjkkalo1_400

wow, this runs much better - tested on the rM1, very cool, thank you!

i think you switched to microsecond to support millisecond resolution timers?

**Comment by raisjn** [comment]
re:namespace: i think ui::MainLoop::set_timeout and such is reasonable (instead of reaching into Timers)

File: src/rmkit/util/timer.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
+#ifndef RMKIT_TIMER_H
```
Comment (raisjn) [inline src/rmkit/util/timer.h:1]: wouldn't this file look so much better as cpy? 🤞🏿

**Comment by mrichards42** [comment]
> i think you switched to microsecond to support millisecond resolution timers?

I switched since timeval uses microsecond resolution, so it seemed like a good idea for consistency. It does make timers more accurate, but given the latency of the rm display I doubt there's any point in using microseconds. I'd be happy to go back to milliseconds (which also means regular ints instead of int64s).

> re:namespace: i think ui::MainLoop::set_timeout and such is reasonable (instead of reaching into Timers)

On a related note -- do you think it's worth having both the set_timeout template (which is supposed to support whichever chrono resolutions you'd like) and set_timeout_ms (taking a long)? It would make the api a lot simpler if there were just one function -- my own preference would be to just have a single set_timeout that uses milliseconds.

File: src/rmkit/util/timer.h (hunk 1 above)
Comment (mrichards42) [reply src/rmkit/util/timer.h:1]: I'm happy to give it a shot! I wrote it in plain c++ just b/c I'm more familiar and I haven't put the time in to come up w/ a decent editor config for cpy (both ft=python and ft=cpp are close, but I haven't found a happy medium where commenting, indentation, and autocomplete do what I want them to). But I imagine the translation to cpy would be pretty straightforward in this case, so definitely don't mind if you prefer that most of rmkit is cpy 🙂

Comment (raisjn) [reply src/rmkit/util/timer.h:1]: commenting and indentation work for me. i don't have autocomplete in general (sorry), just the normal keyword completion. i set ft=cpy and have [cpy.vim](https://gist.github.com/raisjn/1dc8538b151a7e2a1e548128634b9288)

there's also ctags.d/ config in rmkit which you can use with upgraded ctags (universal ctags?) for tag gen / fast jumping between most functions and classes (but not all). also supports tagbar.vim (using context stack), but i don't remember if i had to make changes there. 

Screenshot_2021-02-13_09-48-33

**Comment by raisjn** [comment]
> On a related note -- do you think it's worth having both the set_timeout template (which is supposed to support whichever chrono resolutions you'd like) and set_timeout_ms (taking a long)? It would make the api a lot simpler if there were just one function -- my own preference would be to just have a single set_timeout that uses milliseconds.

one is fine, i also am fine with it being ms (normal case) and going back to int32

**Review by raisjn — APPROVED** [review]
looks good, ran it again on rM1, recompiled all codebase. i like the API, should be familiar to most people


for cpy, you usually don't need surrounding parens on conditionals. 

```
if (a):
  pass
// is equivalent
if a:
  pass
```

and

```
auto foo = bar
foo := bar
```

(i'm not requesting any changes, just noting)

**Comment by mrichards42** [comment]
just doing some testing on puzzles. I'm running into some trouble w/ the "game finished" flash. I think it has something to do w/ the particulars about how and when I was triggering timers before, but will need a little digging to figure out. I don't think it's an issue with this PR, but we'll see.

**Pull request merged**